File: sdk/java/src/main/java/internal/telemetry/SourceMatcher.java
package internal.telemetry;

import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.ArrayList;
import java.util.List;

/**
 * Membership test for a set of CIDR prefixes, used to classify a socket peer.
 *
 * <p>Two properties matter more than speed.
 *
 * <p>First, <strong>no name resolution, ever</strong>. Every candidate is checked
 * against a literal-address shape before it is handed to the platform, so a value that
 * turns out to be a hostname is rejected rather than looked up. A DNS round trip on the
 * request path would be a latency spike that appears only for certain inputs, which is
 * the worst possible shape for something a latency dashboard is built on.
 *
 * <p>Second, a malformed prefix in a deployment variable is skipped, not thrown. A typo
 * in configuration must never stop a service from starting.
 */
public final class SourceMatcher {

    private record Prefix(byte[] network, int bits) {
    }

    private final List<Prefix> prefixes;

    private SourceMatcher(List<Prefix> prefixes) {
        this.prefixes = prefixes;
    }

    /** A matcher that matches nothing. */
    public static SourceMatcher none() {
        return new SourceMatcher(List.of());
    }

    /**
     * Compile a list of prefixes. A bare address is taken as a host route
     * ({@code /32} or {@code /128}).
     */
    public static SourceMatcher of(List<String> entries) {
        List<Prefix> compiled = new ArrayList<>();
        for (String raw : entries) {
            String entry = raw == null ? "" : raw.strip();
            if (entry.isEmpty()) {
                continue;
            }
            int slash = entry.lastIndexOf('/');
            String addressPart = slash < 0 ? entry : entry.substring(0, slash);
            InetAddress address = literal(addressPart);
            if (address == null) {
                continue;
            }
            byte[] bytes = address.getAddress();
            int full = bytes.length * 8;
            int bits = full;
            if (slash >= 0) {
                try {
                    bits = Integer.parseInt(entry.substring(slash + 1).strip());
                } catch (NumberFormatException malformed) {
                    continue;
                }
            }
            if (bits < 0 || bits > full) {
                continue;
            }
            compiled.add(new Prefix(bytes, bits));
        }
        return new SourceMatcher(List.copyOf(compiled));
    }

    public int size() {
        return prefixes.size();
    }

    /** True when {@code candidate} is a literal address inside one of the prefixes. */
    public boolean matches(String candidate) {
        if (prefixes.isEmpty()) {
            return false;
        }
        InetAddress address = literal(normalise(candidate));
        if (address == null) {
            return false;
        }
        byte[] bytes = address.getAddress();
        for (Prefix prefix : prefixes) {
            if (prefix.network().length == bytes.length && contains(prefix, bytes)) {
                return true;
            }
        }
        return false;
    }

    private static boolean contains(Prefix prefix, byte[] address) {
        int fullBytes = prefix.bits() / 8;
        int remainder = prefix.bits() % 8;
        for (int i = 0; i < fullBytes; i++) {
            if (prefix.network()[i] != address[i]) {
                return false;
            }
        }
        if (remainder == 0) {
            return true;
        }
        int mask = 0xff << (8 - remainder);
        return (prefix.network()[fullBytes] & mask) == (address[fullBytes] & mask);
    }

    /**
     * Reduce a socket-reported address to a comparable literal.
     *
     * <p>Handles the four shapes that actually arrive: bracketed IPv6, a zone index
     * ({@code fe80::1%eth0}), a trailing port, and the IPv4-mapped form
     * ({@code ::ffff:10.0.0.4}) that a dual-stack listener reports for an IPv4 peer.
     * Comparing that last form against an IPv4 prefix silently never matches, which is
     * a bug that only shows up in one deployment shape and is invisible in every test
     * that binds to IPv4.
     */
    public static String normalise(String candidate) {
        if (candidate == null) {
            return null;
        }
        String value = candidate.strip();
        if (value.isEmpty()) {
            return null;
        }
        if (value.startsWith("[")) {
            int close = value.indexOf(']');
            value = close > 0 ? value.substring(1, close) : value.substring(1);
        } else if (value.indexOf(':') >= 0 && value.indexOf(':') == value.lastIndexOf(':')) {
            // Exactly one colon: an IPv4 address with a port, never an IPv6 address.
            value = value.substring(0, value.indexOf(':'));
        }
        int zone = value.indexOf('%');
        if (zone > 0) {
            value = value.substring(0, zone);
        }
        String lowered = value.toLowerCase(java.util.Locale.ROOT);
        if (lowered.startsWith("::ffff:") && lowered.indexOf('.') > 0) {
            value = value.substring(7);
        }
        return value;
    }

    /**
     * Parse an address, and only an address.
     *
     * <p>Returns {@code null} for anything that is not a literal, so nothing here can
     * ever reach a resolver.
     */
    public static InetAddress literal(String value) {
        if (value == null || value.isEmpty() || value.length() > 45) {
            return null;
        }
        boolean sawDigit = false;
        for (int i = 0; i < value.length(); i++) {
            char c = value.charAt(i);
            boolean ok = (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F')
                    || c == '.' || c == ':';
            if (!ok) {
                return null;
            }
            sawDigit |= c == '.' || c == ':' || (c >= '0' && c <= '9');
        }
        if (!sawDigit) {
            return null;
        }
        // A bare "abc" would be hex-only and could still be a host name; require the
        // structure of an address.
        if (value.indexOf('.') < 0 && value.indexOf(':') < 0) {
            return null;
        }
        try {
            return InetAddress.getByName(value);
        } catch (UnknownHostException notAnAddress) {
            return null;
        }
    }
}

File: sdk/dotnet/Internal.Telemetry/SourceMatcher.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Net;
using System.Net.Sockets;

namespace Internal.Telemetry;

/// <summary>
/// Address normalisation and prefix matching for the estate's generated-traffic ranges.
/// </summary>
/// <remarks>
/// Membership of those ranges must never be something a caller can claim for itself, so
/// everything here is fed the socket peer address and nothing else. The helpers are
/// public because the middleware, the client and the tests all have to agree on exactly
/// one reading of an address.
/// </remarks>
public sealed class SourceMatcher
{
    private readonly List<Prefix> _prefixes;

    private SourceMatcher(List<Prefix> prefixes)
    {
        _prefixes = prefixes;
    }

    /// <summary>An empty matcher: nothing is ever inside it.</summary>
    public static SourceMatcher Empty { get; } = new SourceMatcher(new List<Prefix>());

    /// <summary>How many prefixes were understood. Unparseable entries are skipped.</summary>
    public int Count => _prefixes.Count;

    /// <summary>
    /// Compile CIDR prefixes. A bare address is taken as a host route (/32 or /128).
    /// A typo in a deployment variable is skipped rather than thrown: it must not stop a
    /// service from starting.
    /// </summary>
    public static SourceMatcher Compile(IEnumerable<string>? prefixes)
    {
        List<Prefix> compiled = new();
        if (prefixes is null)
        {
            return new SourceMatcher(compiled);
        }

        foreach (string raw in prefixes)
        {
            if (string.IsNullOrWhiteSpace(raw))
            {
                continue;
            }

            string entry = raw.Trim();
            int slash = entry.LastIndexOf('/');
            string addressPart = slash < 0 ? entry : entry.Substring(0, slash);
            IPAddress? address = Normalise(addressPart);
            if (address is null)
            {
                continue;
            }

            byte[] bytes = address.GetAddressBytes();
            int full = bytes.Length * 8;
            int bits = full;
            if (slash >= 0)
            {
                string suffix = entry.Substring(slash + 1).Trim();
                if (!int.TryParse(suffix, NumberStyles.Integer, CultureInfo.InvariantCulture, out bits)
                    || bits < 0
                    || bits > full)
                {
                    continue;
                }
            }

            compiled.Add(new Prefix(bytes, bits));
        }

        return new SourceMatcher(compiled);
    }

    /// <summary>True when the address falls inside one of the compiled prefixes.</summary>
    public bool Matches(IPAddress? address)
    {
        if (address is null || _prefixes.Count == 0)
        {
            return false;
        }

        IPAddress folded = Fold(address);
        byte[] candidate = folded.GetAddressBytes();
        foreach (Prefix prefix in _prefixes)
        {
            if (prefix.Contains(candidate))
            {
                return true;
            }
        }

        return false;
    }

    /// <summary>Textual overload, for callers holding an address they read as a string.</summary>
    public bool Matches(string? address) => Matches(Normalise(address));

    /// <summary>
    /// Read an address as a socket might report it.
    /// </summary>
    /// <remarks>
    /// A dual-stack listener reports IPv4 peers as <c>::ffff:10.0.0.4</c>; comparing that
    /// form against an IPv4 prefix silently never matches, so it is folded back first.
    /// Bracketed literals, a trailing port and an interface suffix are all understood,
    /// because every one of them shows up somewhere in a real deployment.
    /// </remarks>
    public static IPAddress? Normalise(string? address)
    {
        if (string.IsNullOrWhiteSpace(address))
        {
            return null;
        }

        string value = address.Trim();

        // [::1]:8080 and [::1]
        if (value.Length > 1 && value[0] == '[')
        {
            int close = value.IndexOf(']');
            if (close > 1)
            {
                value = value.Substring(1, close - 1);
            }
        }
        else if (value.IndexOf(':') > 0 && value.IndexOf(':') == value.LastIndexOf(':'))
        {
            // Exactly one colon: an IPv4 address with a port, never an IPv6 literal.
            value = value.Substring(0, value.IndexOf(':'));
        }

        int zone = value.IndexOf('%');
        if (zone > 0)
        {
            value = value.Substring(0, zone);
        }

        if (!IPAddress.TryParse(value, out IPAddress? parsed))
        {
            return null;
        }

        return Fold(parsed);
    }

    /// <summary>Fold an IPv4-mapped IPv6 address back to its IPv4 form.</summary>
    public static IPAddress Fold(IPAddress address)
    {
        if (address.AddressFamily != AddressFamily.InterNetworkV6)
        {
            return address;
        }

        if (address.IsIPv4MappedToIPv6)
        {
            return address.MapToIPv4();
        }

        // ScopeId is only readable on an IPv6 address, which is why the family is
        // settled above before it is touched. A scoped address is rebuilt without the
        // interface index: no prefix ever carries one.
        return address.ScopeId != 0 ? new IPAddress(address.GetAddressBytes()) : address;
    }

    /// <summary>Render an address the way it travels on a record.</summary>
    public static string? Format(IPAddress? address)
    {
        return address is null ? null : Fold(address).ToString();
    }

    private readonly struct Prefix
    {
        private readonly byte[] _network;
        private readonly int _bits;

        public Prefix(byte[] network, int bits)
        {
            _network = network;
            _bits = bits;
        }

        public bool Contains(byte[] candidate)
        {
            if (candidate.Length != _network.Length)
            {
                return false;
            }

            int wholeBytes = _bits / 8;
            for (int i = 0; i < wholeBytes; i++)
            {
                if (candidate[i] != _network[i])
                {
                    return false;
                }
            }

            int remainder = _bits % 8;
            if (remainder == 0)
            {
                return true;
            }

            int mask = 0xFF << (8 - remainder);
            return (candidate[wholeBytes] & mask) == (_network[wholeBytes] & mask);
        }
    }
}
